bazel/testing/lit_autoupdate_base.py

```diff
@@ -0,0 +1,408 @@
+
+
+class OriginalLine(Line):
+    """A line that was copied from the original test file."""
+
+    def __init__(self, line_number: int, text: str) -> None:
+        self.line_number = line_number
+        self.text = text
+
+    def format(self, **kwargs: Any) -> str:
+        return self.text
+
+
+class CheckLine(Line):
+    """A `// CHECK:` line generated from the test output.
+
+    If there's a line number, it'll be fixed up after we've figured out which
+    lines to include in the resulting test file and in what order, because
+    their contents depend on where an original input line appears in the output.
+    """
+
+    def __init__(
+        self,
+        out_line: str,
+        line_number_format: str,
+        line_number_pattern: Pattern,
+    ) -> None:
+        super().__init__()
+        self.indent = ""
+        self.out_line = out_line.rstrip()
+        self.line_number_format = line_number_format
+        self.line_number_pattern = line_number_pattern
+        self.line_numbers = [
+            int(n) - 1 for n in line_number_pattern.findall(self.out_line)
+        ]
+
+    def format(
+        self, *, output_line_number: int, line_number_remap: Dict[int, int]
+    ) -> str:
+        if not self.out_line:
+            return f"{self.indent}// CHECK-EMPTY:\n"
+        result = self.out_line
+        for line_number in self.line_numbers:
+            delta = line_number_remap[line_number] - output_line_number
+            # We use `:+d` here to produce `LINE-n` or `LINE+n` as appropriate.
+            result = self.line_number_pattern.sub(
+                self.line_number_format % {"delta": f"{delta:+d}"},
+                result,
+                count=1,
+            )
+        return f"{self.indent}// CHECK:{result}\n"
+
+
+class Autoupdate(NamedTuple):
+    line_number: int
+    cmd: str
+
+
+def find_autoupdate(test: str, orig_lines: List[str]) -> Optional[Autoupdate]:
+    """Figures out whether autoupdate should occur.
+
+    For AUTOUPDATE, returns the line and command. For NOAUTOUPDATE, returns
+    None.
+    """
+    found = 0
+    result = None
+    for line_number, line in enumerate(orig_lines):
+        if line.startswith(AUTOUPDATE_MARKER):
+            found += 1
+            result = Autoupdate(line_number, line[len(AUTOUPDATE_MARKER) :])
+        elif line.startswith(NOAUTOUPDATE_MARKER):
+            found += 1
+    if found == 0:
+        raise ValueError(
+            f"{test} must have either '{AUTOUPDATE_MARKER}' or "
+            f"'{NOAUTOUPDATE_MARKER}'"
+        )
+    elif found > 1:
+        raise ValueError(
+            f"{test} must have only one of '{AUTOUPDATE_MARKER}' or "
+            f"'{NOAUTOUPDATE_MARKER}'"
+        )
+    return result
+
+
+def replace_all(s: str, replacements: List[Tuple[str, str]]) -> str:
+    """Runs multiple replacements on a string."""
+    for before, after in replacements:
+        s = s.replace(before, after)
+    return s
+
+
+def get_actual_output(
```

Comment:
Is there a more descriptive name we could use here? It's not clear to me what "actual" means here, especially given that we're doing some post-processing on the output. Would something like `get_matchable_test_output` be reasonable?

Reply:
Done